tasks/inventory_targets.rb

```diff
@@ -0,0 +1,242 @@
+
+    # Filter by location
+    if opts[:location]
+      vms.select! { |vm| vm['location'] == opts[:location] }
+    end
+
+    # Filter by tags - tags are ANDed
+    if opts[:tags]
+      # Tag names are case insensitive, values are case sensitive
+      expected_tags = opts[:tags].map { |name, value| [name.to_s.downcase, value] }.to_h
+
+      vms.select! do |vm|
+        present_tags = vm.fetch('tags', {}).map { |name,value| [name.downcase, value] }.to_h
+        # Hash <= checks whether the lhs is a subset of the rhs
+        expected_tags <= present_tags
+      end
+    end
+
+    vms
+  end
+
+  def nics(token, creds, opts)
+    url = if opts[:resource_group]
+            if opts[:scale_set]
+              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
+              "virtualMachineScaleSets/#{opts[:scale_set]}/" \
+              "networkinterfaces?api-version=2017-03-30"
+            else
+              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Network/" \
+              "networkInterfaces?api-version=2019-07-01"
+            end
+          else
+            "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+            "providers/Microsoft.Network/networkInterfaces?api-version=2019-07-01"
+          end
+
+    get_all_results(url, token)
+  end
+
+  def index_by_id(objects)
+    objects.each_with_object({}) do |obj, acc|
+      acc[obj['id']] = obj
+    end
+  end
+
+  # Uses the client credentials grant flow
+  # https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-oauth2-client-creds-grant-flow
+  def token(creds)
+    data = {
+      grant_type: 'client_credentials',
+      client_id: creds['client_id'],
+      client_secret: creds['client_secret'],
+      resource: 'https://management.azure.com'
+    }
+
+    uri = URI.parse("https://login.microsoftonline.com/#{creds['tenant_id']}/oauth2/token")
+
+    request(:Post, uri, data)
+  end
+
+  def request(verb, uri, data, header = {})
+    # Create the client
+    client = Net::HTTP.new(uri.host, uri.port)
+
+    # Azure REST API always uses SSL
+    client.use_ssl = true
+    client.verify_mode = OpenSSL::SSL::VERIFY_PEER
+
+    # Build the request
+    request = Net::HTTP.const_get(verb).new(uri.request_uri, header)
+
+    # Build the query if there's data to send
+    query = URI.encode_www_form(data) if data
+
+    # Send the request
+    begin
+      response = client.request(request, query)
+    rescue StandardError => e
+      raise TaskHelper::Error.new(
+        "Failed to connect to #{uri}: #{e.message}",
+        'bolt.plugin/azure-http-error'
+      )
+    end
+
+    case response
+    when Net::HTTPOK
+      JSON.parse(response.body)
+    else
+      err = JSON.parse(response.body).dig('error', 'message')
```

Comment:
With bad credentials the task can an error like:
```
Error executing plugin: task from inventory_targets in azure-vms: String does not have #dig method
```
Update to handle different response structure for token request when invalid credentials are supplied. Example error hash below. 
```
Error executing plugin: task from inventory_targets in azure-vms: {"error"=>"invalid_request", "error_description"=>"AADSTS90002: Tenant 'ef17e557-ab6b-4406-b4d4-c28c6991b3c' not found. This may happen if there are no active subscriptions for the tenant. Check with your subscription administrator.\r\nTrace ID: f4bb2961-6aa5-4848-a8d4-9d3989811b00\r\nCorrelation ID: 9a68bb67-07fe-4eba-8002-0a0e68267e1b\r\nTimestamp: 2019-09-04 18:40:05Z", "error_codes"=>[90002], "timestamp"=>"2019-09-04 18:40:05Z", "trace_id"=>"f4bb2961-6aa5-4848-a8d4-9d3989811b00", "correlation_id"=>"9a68bb67-07fe-4eba-8002-0a0e68267e1b", "error_uri"=>"https://login.microsoftonline.com/error?code=90002"}
```

Reply:
Updated to check for `error_description` at the top-level as well as `message` nested under `error`.